third_party/move/move-binary-format/src/file_format.rs

```diff
@@ -1335,503 +1336,1277 @@ pub struct CodeUnit {
-    ///
-    /// Stack transition:
-    ///
-    /// ```..., bool_value -> ..., bool_value```
+
+    #[group = "boolean"]
+    #[description = r#"
+        Invert the bool value at the top of the stack and push the result on the stack.
+    "#]
+    #[semantics = r#"
+        stack >> bool_val
+        stack << (not bool_val)
+    "#]
+    #[runtime_check_epilogue = r#"
+        ty_stack >> ty
+        assert ty == bool
+        ty_stack << bool
+    "#]
     Not,
-    /// Compare for equality the 2 value at the top of the stack and pushes the
-    /// result on the stack.
-    /// The values on the stack must have `Drop` as they will be consumed and destroyed.
-    ///
-    /// Stack transition:
-    ///
-    /// ```..., value(1), value(2) -> ..., bool_value```
+
+    #[group = "comparison"]
+    #[description = r#"
+        Compare for equality the two values at the top of the stack and push the result on the stack.
+
+        The values must have the `drop` ability as they will be consumed and destroyed.
+    "#]
+    #[semantics = r#"
+        stack >> rhs
+        stack >> lhs
+        stack << (lhs == rhs)
+
+        Note that equality is only defined for
+            - Simple primitive types: u8, u16, u32, u64, u128, u256, bool, address
+            - vector<T> where equality is defined for T
+            - &T (or &mut T) where equality is defined for T
+    "#]
+    #[runtime_check_epilogue = r#"
+        ty_stack >> left_ty
+        ty_stack >> right_ty
+        assert left_ty == right_ty
+        assert left_ty has drop
+        ty_stack << bool
+    "#]
     Eq,
-    /// Compare for inequality the 2 value at the top of the stack and pushes the
-    /// result on the stack.
-    /// The values on the stack must have `Drop` as they will be consumed and destroyed.
-    ///
-    /// Stack transition:
-    ///
-    /// ```..., value(1), value(2) -> ..., bool_value```
+
+    #[group = "comparison"]
+    #[description = r#"
+        Similar to `eq`, but with the result being inverted.
+    "#]
+    #[semantics = r#"
+        stack >> rhs
+        stack >> lhs
+        stack << (lhs != rhs)
+    "#]
+    #[runtime_check_epilogue = r#"
+        ty_stack >> left_ty
+        ty_stack >> right_ty
+        assert left_ty == right_ty
+        assert left_ty has drop
+        ty_stack << bool
+    "#]
     Neq,
-    /// Perform a "less than" operation of the 2 u64 at the top of the stack and pushes the
-    /// result on the stack.
-    ///
-    /// Stack transition:
-    ///
-    /// ```..., u64_value(1), u64_value(2) -> ..., bool_value```
+
+    #[group = "comparison"]
+    #[description = r#"
+        Perform a "less than" operation of the two integer values at the top of the stack
+        and push the boolean result on the stack.
+
+        The operands can be of any (but the same) primitive integer type.
+    "#]
+    #[semantics = r#"
+        stack >> (rhs: int_ty)
```

Comment:
In the semantics, we pop off rhs, then lhs. In the runtime_check_epilogue, we pop off left_ty and then right_ty. This happens for several bytecodes and should be fixed.

Reply:
@vgao1996 Just wanted to check if this was fixed.

Reply:
Not yet, but will fix this + check similar instructions before landing.